src/chunk.c

```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
 }
 
 static void
-drop_remote_chunks(FunctionCallInfo fcinfo, const Name schema_name, const Name table_name,
-				   List *data_node_oids)
+lock_referenced_tables(Oid table_relid)
 {
-	/* Wildcard drops not supported on distributed hypertables */
-	if (schema_name == NULL && table_name == NULL)
-		ereport(ERROR,
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("cannot use wildcard to drop chunks on distributed hypertables"),
-				 errhint("Drop chunks on each distributed hypertable individually.")));
-
-	/* The schema name must be present when dropping remote chunks because the
-	 * search path on the connection is always set to pg_catalog. Thus, the
-	 * data node will not be able to resolve the same hypertables without the
-	 * schema. */
-	if (schema_name == NULL)
-		ereport(ERROR,
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("schema required when dropping chunks on distributed hypertables")));
+	List *fk_relids = NIL;
+	ListCell *lf;
+	List *cachedfkeys = NIL;
+	Relation table_rel = table_open(table_relid, AccessShareLock);
 
-	/* Make sure the schema is set in case it was not given on the command line */
-	FC_ARG(fcinfo, 2) = NameGetDatum(schema_name);
-	FC_NULL(fcinfo, 2) = false;
+	/* this list is from the relcache and can disappear with a cache flush, so
+	 * no further catalog access till we save the fk relids */
+	cachedfkeys = RelationGetFKeyList(table_rel);
+	foreach (lf, cachedfkeys)
+	{
+		ForeignKeyCacheInfo *cachedfk = (ForeignKeyCacheInfo *) lfirst(lf);
+
+		/* conrelid should always be that of the table we're considering */
+		Assert(cachedfk->conrelid == RelationGetRelid(table_rel));
+		fk_relids = lappend_oid(fk_relids, cachedfk->confrelid);
+	}
+	table_close(table_rel, AccessShareLock);
 
-	ts_cm_functions->func_call_on_data_nodes(fcinfo, data_node_oids);
+	/* We have a FK between hypertable H and PAR. Hypertable H has number of
+	 * chunks C1, C2, etc. When we execute "drop table C", PG acquires locks
+	 * on C and PAR. If we have a query as "select * from hypertable", this
+	 * acquires a lock on C and PAR as well. But the order of the locks is not
+	 * the same and results in deadlocks. - github issue 865 We hope to
+	 * alleviate the problem by acquiring a lock on PAR before executing the
+	 * drop table stmt. This is not fool-proof as we could have multiple
+	 * fkrelids and the order of lock acquisition for these could differ as
+	 * well. Do not unlock - let the transaction semantics take care of it. */
+	foreach (lf, fk_relids)
+		LockRelationOid(lfirst_oid(lf), AccessExclusiveLock);
+}
+
+/*
+ * Find either the hypertable or the materialized hypertable, if the relid is
+ * a continuous aggregate, for the relid.
+ *
+ * Will error if relid is not a hypertable or view (or cannot be found).
+ *
+ * If `allows_materialization` is false, it will also error if the hypertable
+ * found is a materialized hypertable.
```

Comment:
It means it has to be true to be able to drop chunks directly from the materialization table.

Reply:
> It means it has to be true to be able to drop chunks directly from the materialization table.

What is the purpose? Is it to prevent users from dropping chunks from the materialised hypertable directly? Or is there other cases when it can happen and this prevention is needed?